libr/lang/p/qjs/io.c

```diff
@@ -0,0 +1,258 @@
+		strcpy (buf, "hello");
+	}
+	// read stuff
+	return count;
+#endif
+	RCore *core = io->coreb.core;
+	QjsPluginManager *pm = R_UNWRAP4 (core, lang, session, plugin_data);
+
+	// XXX this logic works for open or check, but not for seek, but its fine for 1 plugin for now
+	ut64 addr = 0; // TODO: oops wheres the current seek
+	QjsIoPlugin *plugin;
+	R_VEC_FOREACH (&pm->io_plugins, plugin) {
+		JSContext *ctx = plugin->ctx;
+		JSValueConst args[2] = {
+			JS_NewInt64 (ctx, addr),
+			JS_NewInt32 (ctx, count),
+		};
+		JSValue res = JS_Call (ctx, plugin->fn_read_js, JS_UNDEFINED, countof (args), args);
+		if (JS_IsArray (ctx, res)) {
+			// JSObject *jo = JS_VALUE_GET_OBJ (res);
+			int len = 10; //eS_ToArrayLengthFree (ctx, &le, res, true);
+			// jint length = JS_GetProperty (ctx, res, JS_ATOM_Array);
+			int i;
+			JSValue jlen = JS_GetPropertyStr(ctx, res, "length");
+			int64_t arrayLength = 10;
+			JS_ToInt64 (ctx, &arrayLength, jlen);
+			int length = R_MIN (arrayLength, count);
+			for (i = 0; i < length; i++) {
+				ut32 u;
+				JSValue element = JS_GetPropertyUint32 (ctx, res, i);
+				if (JS_IsException (element)) {
+					R_LOG_WARN ("exception");
+					break;
+				}
+				if (JS_ToUint32 (ctx, &u, element) != 0) {
+					R_LOG_WARN ("exception2");
+					break;
+				}
+				buf[i] = u;
+			}
+			return len;
+		}
+		R_LOG_WARN ("return is not array");
+		return count;
+	}
+	// read stuff
+	return count;
+}
+
+static ut64 qjs_io_seek(RIO *io, RIODesc *fd, ut64 addr, int whence) {
+	RCore *core = io->coreb.core;
+	QjsPluginManager *pm = R_UNWRAP4 (core, lang, session, plugin_data);
+
+	// XXX this logic works for open or check, but not for seek, but its fine for 1 plugin for now
+	QjsIoPlugin *plugin;
+	R_VEC_FOREACH (&pm->io_plugins, plugin) {
+		JSContext *ctx = plugin->ctx;
+		JSValueConst args[2] = {
+			JS_NewInt64 (ctx, addr),
+			JS_NewInt32 (ctx, whence),
+		};
+		JSValue res = JS_Call (ctx, plugin->fn_seek_js, JS_UNDEFINED, countof (args), args);
+		int64_t radix = 10;
+		JS_ToInt64 (ctx, &radix, res);
+		return (ut64)radix;
+	}
+	return addr;
+}
+
+static RIODesc *qjs_io_open(RIO *io, const char *name, int perm, int mode) {
+	RCore *core = io->coreb.core;
+	QjsPluginManager *pm = R_UNWRAP4 (core, lang, session, plugin_data);
+
+	// Iterate over plugins until one returns "true" (meaning the plugin handled the input)
+	QjsIoPlugin *plugin;
+	R_VEC_FOREACH (&pm->io_plugins, plugin) {
+		JSContext *ctx = plugin->ctx;
+		JSValueConst args[3] = {
+			JS_NewString (ctx, name),
+			JS_NewInt32 (ctx, perm),
+			JS_NewInt32 (ctx, mode),
+	       	};
+		JSValue res = JS_Call (ctx, plugin->fn_open_js, JS_UNDEFINED, countof (args), args);
+		if (JS_ToBool (ctx, res)) {
+			RIOPlugin *iop = plugin->iop;
+			return r_io_desc_new (io, iop, name, perm, mode, core);
+		}
+	}
+
+	return NULL;
+}
+
+// TODO maybe add a function to call by plugin name? (is 1 extra arg)
+static char *qjs_io_system(RIO *io, RIODesc *fd, const char *input) {
```

Comment:
rename to qjs_io_call

Reply:
only if we change the 90 io plugins to use this convention